### PR TITLE
[VEN-2638]: Enable TokenConverters and Prime markets on Ethereum and Sepolia

### PR DESCRIPTION
**Pull request by Debugger022**
## Description

Resolves VEN-2638

This PR contains the vip to enable token converters on Ethereum and Sepolia.

### CODE REVIEW

File: multisig/proposals/sepolia/vip-converter/vipConverter.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      target: WETH_PRIME_CONVERTER,
+      signature: "setConversionConfigs(address,address[],(uint256,uint8)[])",
+      params: [BaseAssets[3], WETHPrimeConverterTokenOuts, incentiveAndAccessibilityForWETHPrimeConverter],
+    },
+    {
+      target: XVS_VAULT_CONVERTER,
+      signature: "setConversionConfigs(address,address[],(uint256,uint8)[])",
+      params: [BaseAssets[4], XVSVaultConverterTokenOuts, incentiveAndAccessibilityForXVSVaultConverter],
+    },
+  ]);
```
Comment (chechu) [inline multisig/proposals/sepolia/vip-converter/vipConverter.ts:93]: I miss the commands to enable the Prime markets. Based on the commands included in the VIP-201, I think these are the missing commands (please, check it):

```
  {
    target: COMPTROLLER_LST,
    signature: "setPrimeToken(address)",
    params: [PRIME],
  },
  {
    target: PRIME_LIQUIDITY_PROVIDER,
    signature: "initializeTokens(address[])",
    params: [[ETH, BTC, USDC, USDT]],
  },
  {
    target: PRIME_LIQUIDITY_PROVIDER,
    signature: "setTokensDistributionSpeed(address[],uint256[])",
    params: [
      [ETH, BTC, USDC, USDT],
      [0, 0, 0, 0],
    ],
  },
  {
    target: PRIME,
    signature: "addMarket(address,uint256,uint256)",
    params: [vETH, ethers.utils.parseEther("2"), ethers.utils.parseEther("4")],
  },
  {
    target: PRIME,
    signature: "addMarket(address,uint256,uint256)",
    params: [vBTC, ethers.utils.parseEther("2"), ethers.utils.parseEther("4")],
  },
  {
    target: PRIME,
    signature: "addMarket(address,uint256,uint256)",
    params: [vUSDC, ethers.utils.parseEther("2"), ethers.utils.parseEther("4")],
  },
  {
    target: PRIME,
    signature: "addMarket(address,uint256,uint256)",
    params: [vUSDT, ethers.utils.parseEther("2"), ethers.utils.parseEther("4")],
  },
  {
    target: PRIME,
    signature: "setLimit(uint256,uint256)",
    params: [
      0, // irrevocable
      500, // revocable
    ],
  },
  {
    target: PRIME_LIQUIDITY_PROVIDER,
    signature: "pauseFundsTransfer()",
    params: [],
  },
```

Comment (Debugger022) [reply multisig/proposals/sepolia/vip-converter/vipConverter.ts:93]: @chechu, the function to add market accepts comptroller as an argument too. So which comptroller and markets, we should consider? Should we add all comptrollers with their markets containing markets from the following ones:
USDC, USDT, WETH, WBTC?

Comment (Debugger022) [reply multisig/proposals/sepolia/vip-converter/vipConverter.ts:93]: we also need to set prime address in **COMPTROLLER_CURVE** and **COMPTROLLER_STABLE_COINS**.

Comment (Debugger022) [reply multisig/proposals/sepolia/vip-converter/vipConverter.ts:156]: Done https://github.com/VenusProtocol/vips/pull/303/commits/7fde183c32cb94c8e7ec50f3c01262679d15c4fe

Comment (Debugger022) [reply multisig/proposals/sepolia/vip-converter/vipConverter.ts:156]: PLP is already paused. So, I have removed that command.

**Comment by Debugger022** [comment]
**Multisig tx Sepolia:** https://sepolia.etherscan.io/tx/0x20962de098e7aafb2c0bf55842d9b4989393827da2c2393e47231a862616026a

File: multisig/proposals/ethereum/vip-converter/Addresses.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  "0x853d955aCEf822Db058eb8505911ED77F175b99e", // FRAX
+  "0xA663B02CF0a4b149d2aD41910CB81e23e1c41c32", // SFRAX
+  "0x6ee2b5E19ECBa773a352E5B21415Dc419A700d1d", // weETH-26DEC2024
+  "0x0000000000085d4780B73119b644AE5ecd22b376", // TUSD
+  "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48", // USDC
+  "0xdAC17F958D2ee523a2206206994597C13D831ec7", // USDT
+  "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599", // WBTC
+  "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2", // WETH
+  "0xCd5fE23C85820F7B72D0926FC9b05b43E359b7ee", // W_E_ETH
+  "0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0", // WST_ETH
```
Comment (chechu) [inline multisig/proposals/ethereum/vip-converter/Addresses.ts:14]: I would add the addresses of rsETH and sfrxETH, that will be added soon:
* rsETH: 0xA1290d69c65A6Fe4DF752f95823fae25cB99e5A7 (https://github.com/VenusProtocol/vips/pull/304)
* sfrxETH: 0xac3E018457B222d93114458476f3E3416Abbe38F (https://github.com/VenusProtocol/vips/pull/302)

File: multisig/proposals/ethereum/vip-converter/vipConverter.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      target: PROTOCOL_SHARE_RESERVE_PROXY,
+      signature: "addOrUpdateDistributionConfigs((uint8,uint16,address)[])",
+      params: [
+        [
+          [0, 6000, VTREASURY],
+          [0, 2000, XVS_VAULT_CONVERTER],
+          [0, 200, USDC_PRIME_CONVERTER],
+          [0, 200, USDT_PRIME_CONVERTER],
+          [0, 100, WBTC_PRIME_CONVERTER],
+          [0, 1500, WETH_PRIME_CONVERTER],
```
Comment (chechu) [inline multisig/proposals/ethereum/vip-converter/vipConverter.ts:78]: According to the [community post](https://community.venus.io/t/venus-prime-deployment-proposal-for-ethereum-mainnet/4417), values should be these:

```suggestion
          [0, 140, USDC_PRIME_CONVERTER], // 7%
          [0, 140, USDT_PRIME_CONVERTER], // 7%
          [0, 140, WBTC_PRIME_CONVERTER], // 7%
          [0, 1580, WETH_PRIME_CONVERTER], // 79%
```

File: multisig/proposals/ethereum/vip-converter/Addresses.ts (hunk 2 above)
Comment (chechu) [reply multisig/proposals/ethereum/vip-converter/Addresses.ts:14]: https://github.com/VenusProtocol/vips/pull/303/commits/d04f88d5fa97464854791a6e5ec1d52425e6d0ed

File: multisig/proposals/ethereum/vip-converter/vipConverter.ts (hunk 3 above)
Comment (chechu) [reply multisig/proposals/ethereum/vip-converter/vipConverter.ts]: https://github.com/VenusProtocol/vips/pull/303/commits/d04f88d5fa97464854791a6e5ec1d52425e6d0ed

File: multisig/simulations/ethereum/vipConverter/index.ts [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+      await pretendExecutingVip(vipConverter());
+
+      converterNetwork = new ethers.Contract(CONVERTER_NETWORK, CONVERTER_NETWORK_ABI, provider);
+      xvsVaultTreasury = new ethers.Contract(XVS_VAULT_TREASURY, XVS_VAULT_CONVERTER_ABI, provider);
+    });
+
+    it("Timelock should be the owner of all converters", async () => {
+      for (const converter of converters) {
+        const Converter = new ethers.Contract(converter, SINGLE_TOKEN_CONVERTER_ABI, provider);
+        expect(await Converter.owner()).to.equal(GUARDIAN);
```
[the remark this answers is not in the document]
Comment (GitGuru7) [reply multisig/simulations/ethereum/vipConverter/index.ts:73]: [ced8de9](https://github.com/VenusProtocol/vips/pull/303/commits/ced8de92fdd5e3f2f737a37702c56f4bc1dc520d)

File: multisig/simulations/ethereum/vipConverter/index.ts [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+const XVS = "0xd3CC9d8f3689B83c91b7B59cAB4946B063EB894A";
+const XVS_VAULT = "0xA0882C2D5DF29233A092d2887A258C2b90e9b994";
+const USDT_HOLDER = "0xb23360CCDd9Ed1b15D45E5d3824Bb409C8D7c460";
+const USDC_HOLDER = "0x974CaA59e49682CdA0AD2bbe82983419A2ECC400";
+
+forking(20120060, () => {
+  const provider = ethers.provider;
+
+  describe("Post-VIP behavior", () => {
```
Comment (GitGuru7) [inline multisig/simulations/ethereum/vipConverter/index.ts:54]: I would add check for new implementation.

File: multisig/simulations/ethereum/vipConverter/index.ts [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    });
+
+    it("Plp should have correct tokens", async () => {
+      expect(await plp.lastAccruedBlockOrSecond(USDT)).to.be.gt(0);
+      expect(await plp.lastAccruedBlockOrSecond(USDC)).to.be.gt(0);
+      expect(await plp.lastAccruedBlockOrSecond(WETH)).to.be.gt(0);
+      expect(await plp.lastAccruedBlockOrSecond(WBTC)).to.be.gt(0);
+    });
+  });
+});
```
Comment (GitGuru7) [inline multisig/simulations/ethereum/vipConverter/index.ts:306]: I will suggest to add some storage layout checks for new implementation of single token converter

Comment (Debugger022) [reply multisig/simulations/ethereum/vipConverter/index.ts:306]: There is not any change in the storage layout, we have only changed the `approve` by `forceApprove`.
Also the tests written regarding conversion and others execute on the new implementation storage. Wdsy?

File: multisig/simulations/ethereum/vipConverter/index.ts (hunk 5 above)
Comment (Debugger022) [reply multisig/simulations/ethereum/vipConverter/index.ts]: Done https://github.com/VenusProtocol/vips/pull/303/commits/52e3de55ff7f4d256ddd6c501e2ab567733ba8cd

**Review by GitGuru7 — APPROVED** [review]

**Review by chechu — APPROVED** [review]

**Comment by chechu** [comment]
https://app.venus.io/#/governance/proposal/326?chainId=56

**Pull request merged**